### ISSUE TITLE
Command computers can be interacted with via keyboards

### ISSUE THREAD

**Issue by SquidDev** [SamsTheNerd/ducky-periphs#28]
## Versions
 - **Minecraft:** Forge 1.20.1
 - **CC: Tweaked:** 1.108.0
 - **Ducky Peripherals:** 1.3.1

## Steps to reproduce
 - Place down a command computer and a keyboard on top of it
 - Switch to survival mode
 - Try to right click on the command computer - the UI will not open, and you get an error message.
 - Right click on the keyboard - it allows you to type on the command computer without problem.

I'm not really sure there's a good way to guard against this, at least within the confines of CC's API. Really the best solution would be to not queue events on command computers, but `IComputerAccess` doesn't expose any methods for that.

Entirely unrelated, but didn't feel worth opening a separate bug for this, `KeyboardPeripheral`'s handling of the computer list isn't thread-safe - `attach` and `detach` can be called from both the computer and main thread. The easiest thing to do is swap out the `ArrayList` with a `Collections.newSetFromMap(new ConcurrentHashMap<>())`.

**Comment by SamsTheNerd** [comment]
is getId() on IComputerAccess the same ID that ServerComputerRegistry.get() takes ? if so could use that to get whether or not it's a command computer ?

Then just check if the player using the keyboard has the right perms (which is OP and creative ? ) and only sending events on the command computers if they have the perms.

**Comment by SquidDev** [comment]
> is getId() on IComputerAccess the same ID that ServerComputerRegistry.get() takes ? if so could use that to get whether or not it's a command computer ?

https://github.com/SamsTheNerd/ducky-periphs/blob/bbf6aaa583fe1f8133ffad946ce9d1b3ac61d340/common/src/main/java/com/samsthenerd/duckyperiphs/peripherals/keyboards/KeyboardPeripheral.java#L92-L94

Wait, what!? I see you were using `IComputerAccess.queueEvent` pre 1.20 - what was the motivation for this switch?

**Comment by SamsTheNerd** [comment]
fixed in 5df3c00036ae3f625b7c08c01a07a9af765f0cf5 & 56252d6ab19f612a15ad9010a6e64661889ea4b0

**Comment by SquidDev** [comment]
> ```java
> ServerComputer computer = ServerContext.get(player.getServer()).registry().get(access.getID());
> ```

I don't think this is correct — `ServerComputerRegistry.get` takes an *instance* ID, which won't be the same as the computer's ID.

Looking at 56252d6ab19f612a15ad9010a6e64661889ea4b0, it might be worth updating to CC:T 1.114, which has a new [`AttachedComputerSet`](https://tweaked.cc/mc-1.20.x/javadoc/dan200/computercraft/api/peripheral/AttachedComputerSet.html) class for making some of this easier. There's some awkward race conditions with detaching peripherals, that a concurrent map/set doesn't handle.

Sorry for coming in and grumbling :p.

**Comment by SamsTheNerd** [comment]
It seems to work 🤷‍♀️ although i suppose it's possible it gets out of wack when you have more computers or something? Any ideas for a better way to do it?

And will do for the concurrency bit 🫡
